docs: :memo: add more context and comparison of frameworks to Why Django post (#56)
seedcase-project/decisions: why-django.qmd

---
title: "Why Django"
description: |

>   Web frameworks are designed to support building both web frontends and API layers. Django is a popular full-stack framework with solutions for most web development tasks out of the box.

Reviewer: For context, here's a copy paste of an earlier discussion of backend vs. back-end (vs. back end) and front end vs front-end :)

> If you believe Meriam-Webster then it is "backend" when it is a noun and "back-end" if it is an adjective, when you are talking computing, and "back end" if anything other than IT. [Definition of BACK END](https://www.merriam-webster.com/dictionary/back%20end).
> And apparently "front end" no matter the context (as a noun) and "front-end" as an adjective.  [Definition of FRONT END](https://www.merriam-webster.com/dictionary/front%20end).

statement in the form of a question for the problem.
:::

> Web applications all have a host of common functionalities (e.g. HTTP request handling, URL routing or data source integration), which take significant time and effort to implement from scratch and maintain for each new project. Good frameworks provide mature, well-maintained components for these features out of the box as well as best-practice patterns and guidelines for structuring code consistently. They also promote code reuse and modular design both within and between applications.

Author: ```suggestion
Web applications all have a host of common functionalities (e.g. HTTP request handling, URL routing or data source integration). These can take significant time and effort to implement from scratch and maintain for each new project. Good frameworks provide mature, well-maintained components for these features out of the box as well as best-practice patterns and guidelines for structuring code consistently. They also promote code reuse and modular design both within and between applications.
```
- Re Oxford comma: unless there's a style rule for always using it, I wouldn't use it here because the list is not super complex
- Re "good": to me it feels like the entire section is about explaining what a good framework is, so I would actually prefer to keep it vague at the beginning and clarify it as we go along

Reviewer 2: In my strong opinion, the Oxford comma should always be used if a list of three or more can't be simplified. And there's strong reasons for using it always e.g. https://thewritelife.com/is-the-oxford-comma-necessary/ we don't follow a specific writing style guide (might be good to consider one). Regardless of the guide though, I will always say to use the serial aka Oxford comma.

> Web applications all have a host of common functionalities (e.g. HTTP request handling, URL routing or data source integration), which take significant time and effort to implement from scratch and maintain for each new project. Good frameworks provide mature, well-maintained components for these features out of the box as well as best-practice patterns and guidelines for structuring code consistently. They also promote code reuse and modular design both within and between applications.

Author: Do you mind if I resolve this based on [my comment](https://github.com/seedcase-project/decisions/pull/56#discussion_r1611270083), which very helpfully disappeared after I committed the corresponding suggestion?

> On the whole, frameworks make applications robust, secure and scaleable, collaborative development quicker and easier, and software maintenance less onerous, allowing the team to focus on implementing project-specific features. 

Author: ```suggestion
Overall, frameworks make applications robust, secure and scaleable, collaborative development quicker and easier, and software maintenance less onerous, allowing the team to focus on implementing project-specific features.
```

> On the whole, frameworks make applications robust, secure and scaleable, collaborative development quicker and easier, and software maintenance less onerous, allowing the team to focus on implementing project-specific features. 

Author: From here I've taken "overall", but I guess I'll keep the rest as is, unless there's a particular reason to change it (which there may be)

> On the whole, frameworks make applications robust, secure and scaleable, collaborative development quicker and easier, and software maintenance less onerous, allowing the team to focus on implementing project-specific features. 

Reviewer: It's fine; it was just a suggestion. It's a personal preference, so if you like what you wrote better, go with that :)

have arisen that impact work.
:::

> We are developing Seedcase in Python (see [Why Python](why-python.qmd) for more information about this decision), so we only consider frameworks for Python. Seedcase is a relatively complex web application with a frontend, an API and a data layer, so we are looking for a full-stack framework that comes with solutions for common tasks in all three of these layers. More particularly, we are looking for tools that facilitate interactive web page design, form handling, URL mapping, serialisation, REST API development and database management. As Seedcase has users with differentiated roles and privileges, a good user management system with support for authentication and authorisation mechanisms is important. Working with relational data, raw data files and image files is a central concern, so the framework should integrate well with popular databases and make it easy to access the local file system.

Author: Resolving, as this was included in a commit

> Additionally, the framework should be mature and well documented, with an active community and extensive resources.

Author: I think compound adjectives in predicative position are not hyphenated

> Additionally, the framework should be mature and well documented, with an active community and extensive resources.

Author: Yeah, documentation, tutorials, forum posts, GitHub issues etc

> Additionally, the framework should be mature and well documented, with an active community and extensive resources.

Author: ```suggestion
Additionally, the framework should be mature and well documented, with an active community and extensive resources (e.g. tutorials, forum posts, GitHub issues).
```

> Additionally, the framework should be mature and well documented, with an active community and extensive resources.

Reviewer: Ah, you're right! Hyphens are used in attributive position, but not often in predicative position.

> Finally, we don't expect an above-average request throughput and are not a real-time data source to external systems, so performance is not the highest ranking factor.

Author: Same subject as in previous clause: we don't expect ... and (we) are not ...

> Finally, we don't expect an above-average request throughput and are not a real-time data source to external systems, so performance is not the highest ranking factor.

Author: Haha, I'll take that

> Finally, we don't expect an above-average request throughput and are not a real-time data source to external systems, so performance is not the highest ranking factor.

Reviewer: @martonvago “(We) are not a real-time data source”?

> Finally, we don't expect an above-average request throughput and are not a real-time data source to external systems, so performance is not the highest ranking factor.

Author: It currently has Luke's "Finally, we don't expect an above-average request throughput **nor are we** a real-time data source...", if that works?

### Django

> [Django](https://www.djangoproject.com) is a full-stack framework for building server-rendered web applications following a Model-View-Template pattern.

Author: ```suggestion
[Django](https://www.djangoproject.com) is a full-stack framework for building server-rendered web applications following a [Model-View-Template](https://www.geeksforgeeks.org/django-project-mvt-structure/) pattern.
```

::: {.columns}
::: {.column}
#### Benefits

> - Has excellent support for API development, including frameworks for REST and GraphQL architectures.

Author: ```suggestion
- Has excellent support for API development, including frameworks for [REST](https://restfulapi.net/) and [GraphQL](https://graphql.org/) architectures.
```

- With the [REST framework](https://www.django-rest-framework.org/) it offers built-in solutions for all common web development tasks, such as HTTP request handling, serialisation, URL routing, pagination, filtering and internationalisation. 
- Comes with a wide range of tools for user management, such as a customisable authentication system with support for social authentication and user profiles, and account-activation and password-reset mechanisms.
- Promotes security best-practices through the built-in authentication system, role-based access control, fine-grained permissions, Cross-Site Request Forgery (CSRF) protection and password hashing.
- Provides a powerful Object-Relational Mapper (ORM) that abstracts away much of the database logic and allows developers to work with database models in a Pythonic way.
- Supports all major databases and has a migration mechanism, making it easy to evolve the database schema over time.
- Has good file storage support, making it easier to work with raw data and image files.

> - Comes with an admin interface out of the box, allowing non-technical users to manage data.

Reviewer 2: ```suggestion
- Comes with an admin interface out of the box, allowing non-technical users to manage the application's data.
```

- Offers built-in form handling capabilities with customisable form classes and form validation logic.
- Uses server-side rendering by default, enabling fast webpage renders. 
- Has a template engine for creating dynamic HTML pages.
- Makes it easy to compose Django applications and has a wide array of well-maintained packages.
- Has excellent documentation and a large and active developer community.
:::
::: {.column}
#### Drawbacks

- Has a steeper learning curve because of its extensive set of features.

> - May be less performant in certain scenarios than the other frameworks considered due to the size of the core framework and its limited support for asynchronous programming.

Author: I kind of think that the actual answer to this is that you can contrive scenarios where Django is outperformed by e.g. FastAPI (e.g. [this benchmark](https://blog.stackademic.com/speed-matters-a-comparative-performance-analysis-of-fastapi-and-django-2d78fd23f909)), but it's hard to say if these tests are representative of real-world use cases and, at any rate, the difference seems to be marginal.

I guess if you had some sort of safety-critical system that handled lots of concurrent, time-consuming incoming and outgoing requests in real time, then you would really care about each ms gained by eliminating idle waiting periods. But then you may not be wanting to develop your safety-critical system as a Python web app…

I initially put this in because every second blog post claims it in some shape or form, but it would probably be better to change the angle a bit and say something like “Django may require more complex configuration and additional extensions to offer the same level of support for concurrency/async programming as the default offered by FastAPI”.

> - May be less performant in certain scenarios than the other frameworks considered due to the size of the core framework and its limited support for asynchronous programming.

Author: Yeah, I also don't know very much about it :sweat_smile: ! My read was that adding async support to Django is an ongoing project: some bits have it (e.g. views), some not yet (e.g. ORM), and for extensions and middleware it varies a lot (https://docs.djangoproject.com/en/5.0/topics/async/). But my understanding is that it is possible to get around these limitations (not that this is terribly relevant for us) by configuring Django in specific ways or swapping sync-only extensions for ones with async support. But one of the selling points of FastAPI is that it is set up for async by default and the whole ecosystem is expected to support it.

> - May be less performant in certain scenarios than the other frameworks considered due to the size of the core framework and its limited support for asynchronous programming.

Author: ```suggestion
- May require more complex configuration and additional extensions to offer the same level of support for asynchronous programming as the default offered by FastAPI.
```
And then I'm not making unqualified and difficult-to-prove claims about performance.

- More opinionated than other frameworks, with less flexibility about the choice of project structure, ORM and template engine.
- Server-rendered websites can have slower response times, which can limit interactivity.

> - Using the built-in template engine creates a tighter coupling between backend and frontend, potentially complicating code reuse and maintenance.

Author: ```suggestion
- Using the built-in template engine creates a tighter coupling between backend and front end, potentially complicating code reuse and maintenance.
```
I really really don't mind either way, but didn't we say "backend" as one word in a computing context above? :sweat_smile:

:::
:::

### Flask

[Flask](https://flask.palletsprojects.com/en/3.0.x/) is a lightweight web framework, designed to be simple and flexible.

::: {.columns}
::: {.column}
#### Benefits

> - Has a gentler learning curve and is easy to use.

Reviewer 2: ```suggestion
- Has a gentler learning curve and might be easier to use compared to other options.
```

- Provides a lot of flexibility without enforcing a particular project structure, ORM or frontend solution.
- Has a smaller footprint, which may lead to a small performance gain.

> - Uses Jinja2 as its default template engine.

Reviewer: This might just be my ignorance, but why's using Jinja2 a benefit? :)

Author: Ah sorry, I meant that the benefit is that is has a default one at all

Reviewer: Ahhh! Maybe write that a bit more explicit then :)

> - Uses Jinja2 as its default template engine.

Author: ```suggestion
- Has a default template engine.
```

- Good option for simple web applications or prototypes.
- Has an active community, good resources and a fair selection of third-party extensions.
:::
::: {.column}
#### Drawbacks

- Provides fewer features out of the box than Django. Authentication, user management, serialisation, form handling, object-relational mapping and the admin interface would need to be implemented from scratch or added through community-maintained extensions.
- Developers have a more active role in responding to security concerns and ensuring that security best practices are implemented appropriately.
:::
:::

### FastAPI

[FastAPI](https://fastapi.tiangolo.com) is a modern web framework for building APIs with a focus on speed and performance. It is particularly well-suited for applications that need to handle a large volume of requests in real time. 

::: {.columns}
::: {.column}
#### Benefits

> - Easy to use and requires minimal setup.

Reviewer 2: ```suggestion
- Easier to use compared to some of the other options and requires less setup.
```

REASONS."
:::

> We decided to choose Django because the requirements of Seedcase are a close match for what Django offers, so we are well-placed to take advantage of many of its core features. Django's built-in ORM, extensive user management capabilities, feature-rich REST framework and solid security principles will save us significant development time and effort and help us create a reliable and long-lived product. The native template engine is also attractive, in that it will allow us to iterate over user interfaces quickly, without the overhead of maintaining a separate frontend framework.

Reviewer: ```suggestion
We decided on Django because the requirements of Seedcase are a close match for what Django offers, so we are well-placed to take advantage of many of its core features. Django's built-in ORM, extensive user management capabilities, feature-rich REST framework, and solid security principles will save us significant development time and effort and help us create a reliable and long-lived product. The native template engine is also attractive, in that it will allow us to iterate over user interfaces quickly, without the overhead of maintaining a separate front-end framework.
```

> 
> 

Reviewer: ```suggestion

```

> While Flask and FastAPI offer more flexibility and autonomy, this is not necessarily an unambiguous positive for us, because we would need to make and justify more decisions about our choice of project structure and third-party extensions. Moreover, the more extensions we rely on, the greater the effort involved in vetting for security and managing dependency updates. As Seedcase is not a performance-critical application, the performance gain offered by FastAPI is not a particularly relevant factor.

Reviewer 2: ```suggestion
While Flask and FastAPI offer more flexibility and autonomy, this is not necessarily an unambiguous positive for us, because we would need to make and justify more decisions about our choice of project structure and third-party extensions. Moreover, the more extensions we rely on, the greater the effort involved in vetting for security and managing dependency updates. As Seedcase software products will likely not be performance-critical applications, the performance gain offered by FastAPI is not a particularly relevant factor.
```

> 
> 

Reviewer: ```suggestion

```

### Consequences

::: content-hidden
List some potential consequences of this decision.
:::

As we will be rendering our web pages on the server side, we should keep an eye out for any user-experience issues in our interaction-heavy user flows, like the data upload or the data project application process. We may consider streamlining interactions on the client side using inline JavaScript in our Django templates.

> A further point to keep in mind is that the integrated, template-based architecture may create a tighter coupling between the frontend and the API layer. While we don't have a specific need to decouple the frontend more formally — as we will not have other user interfaces linked to the API and external interacting systems are well-served by dedicated API endpoints — tightly coupled components can become difficult to refactor or reuse as the application's complexity grows.

Reviewer: ```suggestion
A further point to keep in mind is that the integrated template-based architecture may create a tighter coupling between the front end and the API layer. While we don't have a specific need to decouple the frontend more formally — as we will not have other user interfaces linked to the API and external interacting systems are well-served by dedicated API endpoints — tightly coupled components can become difficult to refactor or reuse as application complexity grows.

```
